bitcoin/src/consensus/serde.rs

```diff
@@ -443,8 +444,13 @@ impl<E: fmt::Debug, I: Iterator<Item = Result<u8, E>>> IterReader<E, I> {
 }
 
 impl<E: fmt::Debug, I: Iterator<Item = Result<u8, E>>> Read for IterReader<E, I> {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+    fn read(&mut self, mut buf: &mut [u8]) -> io::Result<usize> {
         let mut count = 0;
+        if let Some(first) = self.buf.take() {
```

Comment:
Looks like you've removed the if instead of moving it...

Reply:
I'm not sure what you were expecting but ..

We have the `BufRead` implementation reading into a single byte buffer and the `Read` implementation reading as much of the iterator as will fit in the `buf` argument. That seems surprising. Wouldn't it be better to use some arbitrary buffer 1024 bytes or some thing? 

(My fail again for not thinking hard enough during when doing the original PR.)

Reply:
The `IterReader` only reads and buffers a single byte at a time because that's what we get from an iterator. I agree it's a weird API and I'd need to reread the PR to remember what's going on here.

But that has nothing to do with the code at hand, which used to check whether `buf` was empty (and in this case it would do nothing), and now panics instead.

Reply:
The `IterReader` only reads and buffers a single byte at a time because that's what we get from an iterator. I agree it's a weird API and I'd need to reread the PR to remember what's going on here.

But that has nothing to do with the code at hand, which used to check whether `buf` was empty (and in this case it would do nothing), and now panics instead.

Reply:
Exactly, the issue is panic when an empty slice is provided (the type really should've been `NonEmptyOutBuf` but it's too late for that).

Regarding buffer length, since the input is already buffered and the hex decoding doesn't involve context switch or dynamic dispatch I assume it wouldn't help performance but quite likely make it worse. Maybe unless it prevents SIMD or something. If anyone has a benchmark proving otherwise I'd be happy to accept a change but I don't consider it a priority to write it myself.